alloydb-jdbc-connector/src/main/java/com/google/cloud/alloydb/Refresher.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.alloydb;
+
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/** Handles periodic refresh operations for an instance. */
+class Refresher {
+  private static final Logger logger = Logger.getLogger(Refresher.class.getName());
+
+  private final ListeningScheduledExecutorService executor;
+
+  private final Object instanceDataGuard = new Object();
+  private final AsyncRateLimiter rateLimiter;
+
+  private final RefreshCalculator refreshCalculator = new RefreshCalculator();
+  private final Supplier<ListenableFuture<ConnectionInfo>> refreshOperation;
+  private final String name;
+
+  @GuardedBy("instanceDataGuard")
```

Comment:
Let's replace all instanceData references with ConnectionInfo.

Reply:
I'll submit a separate PR for this.